### PR TITLE
feat: (Geospatial) Add support for ReverseGeocoder and CheckPointInPolygon

### PR DESCRIPTION
**Pull request by SudhindraKovalam**
- Added Support for Get Point in Polygon and BatchReverseGeocoding APIs
- Added a few of Spatial schema objects (More to be added with corresponding apis)
- Added a HasSetGeography trait to support creator's geographical endpoints
- Added HasUDID, HasLatLonPair traits
- Refactored HasAddressInput to separate HasAddressInput and BatchAddressGeocoding traits
- Added BatchReverseAddressGeocoding trait to support Batch Reverse geocoding
- Renamed AzureMapsSearchSuite to AzureMapsSuite in tests folder

### CODE REVIEW

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala [hunk 1]
```diff
@@ -19,11 +21,65 @@ import java.net.{URI, URLEncoder}
 import java.util.concurrent.TimeoutException
 import scala.concurrent.blocking
 import scala.language.{existentials, postfixOps}
-import spray.json.DefaultJsonProtocol.{StringJsonFormat, seqFormat}
+import spray.json.DefaultJsonProtocol.{DoubleJsonFormat, StringJsonFormat, seqFormat}
+
+trait HasSetGeography extends Wrappable with HasURL with HasUrlPath {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:26]: Is this needed? Can we just use setLocation?

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+   "source": [
+    "import os\n",
+    "import requests\n",
+    "\n",
+    "if os.environ.get(\"AZURE_SERVICE\", None) == \"Microsoft.ProjectArcadia\":\n",
+    "    from pyspark.sql import SparkSession\n",
+    "    spark = SparkSession.builder.getOrCreate()\n",
+    "\n",
+    "# Azure Maps account key\n",
+    "azureMapsKey = \"YOUR_AZURE_MAPS_KEY_GOES_HERE\"\n",
```
Comment (mhamilton723) [inline notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb:40]: we will need to add the key to the build, contact me on IM

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def extract_point_in_polygon_result_fields(df):\n",
+    "    # Use this function to select only lat/lon columns into the dataframe\n",
+    "    return df.select(\n",
```
Comment (mhamilton723) [inline notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb:152]: could make this shorter with a 
```
.select(col(*), col("output.result.pointInPolygons"), col("output.result.intersectingGeometries"))
.drop("output")
```

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    "    .setUserDataIdentifier(user_data_id)\n",
+    "    .setLatitudeCol(\"lat\")\n",
+    "    .setLongitudeCol(\"lon\")\n",
+    "    .setOutputCol(\"output\"))\n",
+    "\n",
+    "\n",
+    "flood_plain_addresses = extract_point_in_polygon_result_fields(check_point_in_polygon.transform(geocoded_addresses)).collect()\n",
+    "\n",
+    "# Display the results\n",
+    "flood_plain_addresses.show(5)"
```
Comment (mhamilton723) [inline notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb:178]: change to display(...) for nicer results

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    "    .setSubscriptionKey(azureMapsKey)\n",
+    "    .setAddressCol(\"FullAddress\")\n",
+    "    .setOutputCol(\"output\"))\n",
+    "\n",
+    "# Set up a fixed mini batch transformer to geocode addresses\n",
+    "batched_dataframe = geocoder.transform(FixedMiniBatchTransformer().setBatchSize(10).transform(subset_data.coalesce(1)))\n",
+    "geocoded_addresses = extract_location_fields(FlattenBatch().transform(batched_dataframe)).collect()\n",
+    "\n",
+    "# Display the results\n",
+    "geocoded_addresses.show(5)"
```
Comment (mhamilton723) [inline notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb:132]: df.show() -> display(df)

File: notebooks/features/cognitive_services/GeospatialServices - Overview.ipynb [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    "                                47.67856488312544\n",
+    "                            ]\n",
+    "                        ]\n",
+    "                    ]\n",
+    "                } \n",
+    "            } \n",
+    "        ] \n",
+    "    })\n",
+    "\n",
+    "longRunningOperation = r.headers.get('location')\n",
```
Comment (mhamilton723) [inline notebooks/features/cognitive_services/GeospatialServices - Overview.ipynb:232]: nit: should be snake case here

**Review by mhamilton723 — CHANGES_REQUESTED** [review]
Just had time for a quick check but on the whole it looks quite nice! Added a few comments and nits

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala [hunk 7]
```diff
@@ -64,6 +124,72 @@ trait HasAddressInput extends HasServiceParams with HasSubscriptionKey with HasU
   }
 }
 
+trait BatchReverseAddressGeocoding extends HasServiceParams
+  with HasSubscriptionKey with HasURL with HasLatLonPairInput{
+  protected def prepareEntity: Row => Option[AbstractHttpEntity]
+
+  protected def contentType: Row => String = { _ => "application/json" }
+
+  protected def inputFunc(schema: StructType): Row => Option[HttpRequestBase] = {
```
Comment (mhamilton723) [inline cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:133]: I see you override this function as opposed to prepareEntity and prepareEntity doesn't have an impl. Is this your desired behavior? Seems like you might be able to remove code by just overriding prepareEntity instead

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala (hunk 1 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:26]: Azure Maps creator is a geographic resource, although it does the same thing as `HasSetLocation`. This trait was created to reduce confusion

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala (hunk 7 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:133]: Will ping on IM to better understand this

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb (hunk 3 above)
Comment (SudhindraKovalam) [reply notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb]: done!

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb (hunk 5 above)
Comment (SudhindraKovalam) [reply notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb]: done

File: notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb (hunk 4 above)
Comment (SudhindraKovalam) [reply notebooks/features/cognitive_services/GeospatialServices - Flooding Risk.ipynb]: done

File: notebooks/features/cognitive_services/GeospatialServices - Overview.ipynb (hunk 6 above)
Comment (SudhindraKovalam) [reply notebooks/features/cognitive_services/GeospatialServices - Overview.ipynb]: done

File: cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala (hunk 7 above)
Comment (SudhindraKovalam) [reply cognitive/src/main/scala/com/microsoft/azure/synapse/ml/geospatial/AzureMapsHelpers.scala:133]: As discussed. Removed `PrepareEntity` and `ContentType` for now. Will circle back to refactor `CognitiveServicesInput` to support both auth mechanisms

**Comment by mhamilton723** [comment]
COntinuing this PR in #1339

**Pull request closed without merging**